### PR TITLE
Implementation of the Shifted Root-NormLhalf

### PR DESCRIPTION
**Pull request by rjbaraldi**
Works with the BPDN example, LM example fails because of the ODE solver failing.

### CODE REVIEW

File: src/shiftedRootNormLhalf.jl [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+shifted(h::RootNormLhalf{R}, xk::AbstractVector{R}) where {R <: Real} =
+  ShiftedRootNormLhalf(h, xk, zero(xk), false)
+shifted(
+  ψ::ShiftedRootNormLhalf{R, V0, V1, V2},
+  sj::AbstractVector{R},
+) where {R <: Real, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}} =
+  ShiftedRootNormLhalf(ψ.h, ψ.xk, sj, true)
+
+fun_name(ψ::ShiftedRootNormLhalf) = "shifted Lhalf norm"
```
Comment (dpo) [inline src/shiftedRootNormLhalf.jl:34]:
```suggestion
fun_name(ψ::ShiftedRootNormLhalf) = "shifted L½  norm"
```

File: src/normLhalf.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+RootNormLhalf(lambda::R=1) where {R <: Real} = RootNormLhalf{R}(lambda)
+
+function (f::RootNormLhalf)(x::AbstractArray{T}) where {T <: Real}
+  return f.lambda * T(sum(sqrt.(abs.(x))))
+end
+
+function prox!(y::AbstractArray{T}, f::RootNormLhalf, x::AbstractArray{T}, gamma::Real=1) where {T <: Real}
+  γλ = 2 * gamma * f.lambda
+  ϕ(z) = acos(γλ / 8 * (abs(z) /3 )^(-3/2))
```
Comment (dpo) [inline src/normLhalf.jl:33]:
```suggestion
  ϕ(z) = acos(2γλ / 8 * (abs(z) / 3 )^(-3/2))
```
Just so it's a bit more explicit.

File: src/normLhalf.jl [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  for i in eachindex(x)
+    if abs(x[i]) <= 54^(1/3) * (γλ^(2/3)) / 4
```
Comment (dpo) [inline src/normLhalf.jl:35]:
```suggestion
    if abs(x[i]) <= 54^(1/3) * (2γλ)^(2/3) / 4
```

File: src/normLhalf.jl [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+end
+
+RootNormLhalf(lambda::R=1) where {R <: Real} = RootNormLhalf{R}(lambda)
+
+function (f::RootNormLhalf)(x::AbstractArray{T}) where {T <: Real}
+  return f.lambda * T(sum(sqrt.(abs.(x))))
+end
+
+function prox!(y::AbstractArray{T}, f::RootNormLhalf, x::AbstractArray{T}, gamma::Real=1) where {T <: Real}
+  γλ = 2 * gamma * f.lambda
```
Comment (dpo) [inline src/normLhalf.jl:32]:
```suggestion
  γλ = gamma * f.lambda
```

File: src/normLhalf.jl [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      y[i] = 2 * sign(x[i]) / 3 * abs(x[i]) * (1 + cos(2 * π / 3 - 2 * ϕ(x[i]) / 3))
+    end
+  end
+
+  return f.lambda * sum(sqrt.(abs.(y)))
+end
+
+fun_name(f::RootNormLhalf) = "L_(1/2)^(1/2) pseudo-norm"
+fun_dom(f::RootNormLhalf) = "AbstractArray{Real}, AbstractArray{Complex}"
+fun_expr(f::RootNormLhalf{T}) where {T <: Real} = "x ↦ (λ/2)||x||^(1/2)_(1/2)"
```
Comment (dpo) [inline src/normLhalf.jl:47]:
```suggestion
fun_expr(f::RootNormLhalf{T}) where {T <: Real} = "x ↦ ½ λ ‖x‖_(½)^(½)"
```

File: src/normLhalf.jl [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+fun_params(f::RootNormLhalf{T}) where {T <: Real} = "λ = $(f.lambda)"
+
+function prox_naive(f::RootNormLhalf, x::AbstractArray{T}, gamma::Real=1) where {R, T <: Real}
```
Comment (dpo) [inline src/normLhalf.jl:50]: There already is a default `prox(f, x)`, isn't there?

File: src/normLhalf.jl [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      y[i] = 0
+    else
+      y[i] = 2 * sign(x[i]) / 3 * abs(x[i]) * (1 + cos(2 * π / 3 - 2 * ϕ(x[i]) / 3))
+    end
+  end
+
+  return f.lambda * sum(sqrt.(abs.(y)))
+end
+
+fun_name(f::RootNormLhalf) = "L_(1/2)^(1/2) pseudo-norm"
```
Comment (dpo) [inline src/normLhalf.jl:45]:
```suggestion
fun_name(f::RootNormLhalf) = "L½^(½) pseudo-norm"
```

File: src/shiftedRootNormLhalf.jl [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+fun_expr(ψ::ShiftedRootNormLhalf) = "t ↦ ‖xk + sk + t‖ₚᵖ, p = 1/2"
+fun_params(ψ::ShiftedRootNormLhalf) = "xk = $(ψ.xk)\n" * " "^14 * "sj = $(ψ.sj)"
+
+function prox!(
+  y::AbstractVector{R},
+  ψ::ShiftedRootNormLhalf{R, V0, V1, V2},
+  q::AbstractVector{R},
+  σ::R,
+) where {R <: Real, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}}
+  γλ = 2 * σ * ψ.λ
```
Comment (dpo) [inline src/shiftedRootNormLhalf.jl:44]: same suggestion as above

**Review by dpo — COMMENTED** [review]
Thanks! Just a few minor things.

File: src/normLhalf.jl (hunk 5 above)
Comment (rjbaraldi) [reply src/normLhalf.jl:47]: Nice

File: src/normLhalf.jl (hunk 6 above)
Comment (rjbaraldi) [reply src/normLhalf.jl:50]: Yes; for some reason a lot of ProximalOperators have a prox_naive(). I kept conventions here but don't see why they have this.

File: src/shiftedRootNormLhalf.jl (hunk 8 above)
Comment (rjbaraldi) [reply src/shiftedRootNormLhalf.jl:44]: to do 2*gamma*lambda?

File: src/shiftedRootNormLhalf.jl [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  ϕ(z) = acos(γλ / 8 * (abs(z) /3 )^(-3/2))
```
Comment (rjbaraldi) [inline src/shiftedRootNormLhalf.jl:45]:
```suggestion
  ϕ(z) = acos(2γλ / 8 * (abs(z) /3 )^(-3/2))
```

File: src/shiftedRootNormLhalf.jl [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  p = 54^(1/3) * (γλ^(2/3)) / 4
```
Comment (rjbaraldi) [inline src/shiftedRootNormLhalf.jl:46]:
```suggestion
  p = 54^(1/3) * ((2γλ)^(2/3) / 4
```

File: src/normLhalf.jl (hunk 6 above)
Comment (dpo) [reply src/normLhalf.jl:50]: I think it's a straightforward, but potentially inefficient, implementation. Probably used to validate results and for unit tests.

File: src/shiftedRootNormLhalf.jl (hunk 8 above)
Comment (dpo) [reply src/shiftedRootNormLhalf.jl:44]: Yes.

File: src/shiftedRootNormLhalf.jl (hunk 9 above)
Comment (dpo) [reply src/shiftedRootNormLhalf.jl:45]: Any reason why this one isn't `γλ / 4`?

Comment (rjbaraldi) [reply src/shiftedRootNormLhalf.jl:45]: It could be. If γλ = 2 * σ * ψ.λ then it isn't, but if it doesn't contain the 2*() then it can be 4.

**Comment by dpo** [comment]
I hope I rebased this correctly.

**Comment by dpo** [comment]
apparently not :-(

**Comment by dpo** [comment]
Could you confirm that this looks good and merge if it does?

File: src/rootNormLhalf.jl [hunk 11]
```diff
@@ -29,11 +29,11 @@ function (f::RootNormLhalf)(x::AbstractArray{T}) where {T <: Real}
...
 
 function prox!(y::AbstractArray{T}, f::RootNormLhalf, x::AbstractArray{T}, gamma::Real=1) where {T <: Real}
-  γλ = 2 * gamma * f.lambda
-  ϕ(z) = acos(γλ / 8 * (abs(z) /3 )^(-3/2))
+  γλ = gamma * f.lambda
+  ϕ(z) = acos(γλ / 4 * (abs(z) /3 )^(-3/2))
   ysum = zero(T)
   for i in eachindex(x)
-    if abs(x[i]) <= 54^(1/3) * (γλ^(2/3)) / 4
+    if abs(x[i]) <= 54^(1/3) * ((2γλ)^(2/3)) / 4
```
Comment (rjbaraldi) [inline src/rootNormLhalf.jl:36]: If we are really being stingy here, would this add a another multiplication every loop here?

File: src/rootNormLhalf.jl [hunk 12]
```diff
@@ -29,11 +29,11 @@ function (f::RootNormLhalf)(x::AbstractArray{T}) where {T <: Real}
 end
 
 function prox!(y::AbstractArray{T}, f::RootNormLhalf, x::AbstractArray{T}, gamma::Real=1) where {T <: Real}
-  γλ = 2 * gamma * f.lambda
-  ϕ(z) = acos(γλ / 8 * (abs(z) /3 )^(-3/2))
+  γλ = gamma * f.lambda
+  ϕ(z) = acos(γλ / 4 * (abs(z) /3 )^(-3/2))
   ysum = zero(T)
   for i in eachindex(x)
```
Comment (dpo) [inline src/rootNormLhalf.jl:35]:
```suggestion
  threshold = 54^(1/3) * ((2γλ)^(2/3)) / 4
  for i in eachindex(x)
```

File: src/rootNormLhalf.jl (hunk 11 above)
Comment (dpo) [inline src/rootNormLhalf.jl:36]:
```suggestion
    if abs(x[i]) <= threshold
```

Comment (dpo) [reply src/rootNormLhalf.jl:36]: Yes but the entire rhs is reevaluated at each pass. That's a good comment. I'll factor it out.

**Pull request merged**